backend/src/main/java/pt/ulisboa/tecnico/socialsoftware/tutor/submission/SubmissionService.java

```diff
@@ -0,0 +1,219 @@
+
+        if(user.isStudent() && submission.getQuestion().getStatus() != Question.Status.IN_REVISION) {
+            throw new TutorException(CANNOT_EDIT_REVIEWED_QUESTION);
+        }
+
+        this.questionService.updateQuestion(submissionDto.getQuestion().getId(), submissionDto.getQuestion());
+        return new SubmissionDto(submission);
+    }
+
+    @Retryable(value = {SQLException.class}, backoff = @Backoff(delay = 5000))
+    @Transactional(isolation = Isolation.REPEATABLE_READ)
+    public void toggleInReviewStatus(int questionId, boolean inReview) {
+        String status = inReview ? "IN_REVIEW" : "IN_REVISION";
+        updateQuestionStatus(status, questionId);
+    }
+
+    @Retryable(value = { SQLException.class }, backoff = @Backoff(delay = 5000))
+    @Transactional(isolation = Isolation.REPEATABLE_READ)
+    public List<SubmissionDto> getStudentSubmissions(Integer studentId, Integer courseExecutionId) {
+        return submissionRepository.getSubmissions(studentId, courseExecutionId).stream().map(SubmissionDto::new)
+                .collect(Collectors.toList());
+    }
+
+    @Retryable(value = { SQLException.class }, backoff = @Backoff(delay = 5000))
+    @Transactional(isolation = Isolation.REPEATABLE_READ)
+    public List<SubmissionDto> getCourseExecutionSubmissions(Integer courseExecutionId) {
+        return submissionRepository.getCourseExecutionSubmissions(courseExecutionId).stream().map(SubmissionDto::new)
+                .collect(Collectors.toList());
+    }
+
+    @Retryable(value = { SQLException.class }, backoff = @Backoff(delay = 5000))
+    @Transactional(isolation = Isolation.REPEATABLE_READ)
+    public List<ReviewDto> getSubmissionReviews(Integer submissionId) {
+        return reviewRepository.getSubmissionReviews(submissionId).stream().map(ReviewDto::new)
+                .collect(Collectors.toList());
+    }
+
+    @Retryable(value = { SQLException.class }, backoff = @Backoff(delay = 5000))
+    @Transactional(isolation = Isolation.REPEATABLE_READ)
+    public List<UserSubmissionInfoDto> getAllStudentsSubmissionsInfo(Integer courseExecutionId) {
+        Map<Integer, UserSubmissionInfoDto> userSubmissionInfoDtos = new HashMap<>();
+        List<Submission> submissions = submissionRepository.getCourseExecutionSubmissions(courseExecutionId);
+
+        for (Submission submission: submissions) {
+            User user = submission.getUser();
+            if (userSubmissionInfoDtos.containsKey(user.getId())) {
+                userSubmissionInfoDtos.get(user.getId()).addSubmission();
+            } else {
+                userSubmissionInfoDtos.put(user.getId(), new UserSubmissionInfoDto(user, 1));
+            }
+        }
+
+        return new ArrayList<>(userSubmissionInfoDtos.values());
+    }
+
+    private void checkIfConsistentSubmission(SubmissionDto submissionDto) {
```

Comment:
not sure but guessing whether some of these checks can be done inside the domain classes...